### PR TITLE
Fix class memaccess

### LINKED ISSUE #24417: -Wclass-memaccess warnings compiling with GCC 8
**Godot version:**
Master (4b6e7b5)

**OS/device including version:**
Linux, GCC 8.

**Issue description:**
Those two warnings are raised when compiling with `scons p=x11 dev=yes` (`tools=yes target=debug`):
```
In file included from ./core/dvector.h:34,
                 from ./core/variant.h:41,
                 from ./core/method_ptrcall.h:36,
                 from ./core/method_bind.h:35,
                 from ./core/class_db.h:34,
                 from ./core/resource.h:34,
                 from ./core/io/resource_loader.h:34,
                 from modules/pvr/texture_loader_pvr.h:34,
                 from modules/pvr/texture_loader_pvr.cpp:31:
modules/pvr/texture_loader_pvr.cpp: In function 'void _compress_pvrtc4(Image*)':
./core/os/copymem.h:44:56: warning: 'void* memcpy(void*, const void*, size_t)' writing to an object of non-trivially copyable type 'class Javelin::ColorRgba<unsigned char>'; use copy-assignment or copy-initialization instead [-Wclass-memaccess]
 #define copymem(to, from, count) memcpy(to, from, count)
                                                        ^
modules/pvr/texture_loader_pvr.cpp:218:4: note: in expansion of macro 'copymem'
    copymem(bm.GetData(), &r[ofs], size);
    ^~~~~~~
In file included from thirdparty/pvrtccompressor/PvrTcEncoder.h:4,
                 from modules/pvr/texture_loader_pvr.cpp:32:
thirdparty/pvrtccompressor/ColorRgba.h:81:7: note: 'class Javelin::ColorRgba<unsigned char>' declared here
 class ColorRgba : public ColorRgb<T> {
       ^~~~~~~~~
```
```
In file included from ./core/dvector.h:34,
                 from ./core/image.h:35,
                 from ./servers/visual_server.h:34,
                 from ./servers/visual/rasterizer.h:35,
                 from servers/visual/visual_server_scene.h:34,
                 from servers/visual/visual_server_scene.cpp:31:
servers/visual/visual_server_scene.cpp: In member function 'void VisualServerScene::_update_instance_lightmap_captures(VisualServerScene::Instance*)':
./core/os/copymem.h:45:47: warning: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'struct Color'; use assignment or value-initialization instead [-Wclass-memaccess]
 #define zeromem(to, count) memset(to, 0, count)
                                               ^
servers/visual/visual_server_scene.cpp:1243:2: note: in expansion of macro 'zeromem'
  zeromem(p_instance->lightmap_capture_data.ptrw(), 12 * sizeof(Color));
  ^~~~~~~
In file included from ./core/image.h:34,
                 from ./servers/visual_server.h:34,
                 from ./servers/visual/rasterizer.h:35,
                 from servers/visual/visual_server_scene.h:34,
                 from servers/visual/visual_server_scene.cpp:31:
./core/color.h:39:8: note: 'struct Color' declared here
 struct Color {
        ^~~~~
```

Input from @reduz:
```
11:24 <reduz> first one no idea, second one seems maybe clearing manually may be better (or not using color)
```

### PR DESCRIPTION
**Pull request by marxin**
May I please ask for a testing of the changes? I can't verify that.

*Bugsquad edit:* Fixes #24417.

### CODE REVIEW

File: servers/visual/visual_server_scene.cpp [hunk 1]
```diff
@@ -1240,7 +1241,9 @@ void VisualServerScene::_update_instance_lightmap_captures(Instance *p_instance)
 
 	//print_line("update captures for pos: " + p_instance->transform.origin);
 
-	zeromem(p_instance->lightmap_capture_data.ptrw(), 12 * sizeof(Color));
+  for (int i = 0; i < 12; i++)
```
Comment (akien-mga) [inline servers/visual/visual_server_scene.cpp]: Indentation issue, should use tabs and not 2 spaces.

File: modules/pvr/texture_loader_pvr.cpp [hunk 2]
```diff
@@ -215,12 +216,10 @@ static void _compress_pvrtc4(Image *p_img) {
 			int ofs, size, w, h;
 			img->get_mipmap_offset_size_and_dimensions(i, ofs, size, w, h);
 			Javelin::RgbaBitmap bm(w, h);
-			copymem(bm.GetData(), &r[ofs], size);
-			{
+			for (unsigned i = 0; i < size / 4; i++) {
```
Comment (akien-mga) [inline modules/pvr/texture_loader_pvr.cpp]: Should use another letter, this is already in a `for (int i)` loop.

Comment (marxin) [reply modules/pvr/texture_loader_pvr.cpp]: Stupid mistake :+1: I would expect we as GCC can produce a warning for it..

Comment (marxin) [reply modules/pvr/texture_loader_pvr.cpp]: It does with ```-Wshadow```

Comment (akien-mga) [reply modules/pvr/texture_loader_pvr.cpp]: You do, that's how I found about it :)
https://travis-ci.org/godotengine/godot/jobs/484316612

**Comment by akien-mga** [comment]
I'm not familiar with it myself, but we have `memnew_placement` defined in `core/os/memory.h`, and I believe it might be the right one to use here. We have no use of `<new>` in the codebase currently (excluding `thirdparty/`).

**Comment by marxin** [comment]
> I'm not familiar with it myself, but we have `memnew_placement` defined in `core/os/memory.h`, and I believe it might be the right one to use here. We have no use of `<new>` in the codebase currently (excluding `thirdparty/`).

Ah, good. Apparently one patch can utilize ```memnew_arr```, but for the second one one would need an alternative of ```memnew_placement``` that would accept varargs (...) as ctor arguments.

**Comment by akien-mga** [comment]
Right, it's probably simpler to use `new` in the case where you do need varargs then.

File: modules/pvr/texture_loader_pvr.cpp [hunk 3]
```diff
@@ -215,12 +216,10 @@ static void _compress_pvrtc4(Image *p_img) {
...
 			img->get_mipmap_offset_size_and_dimensions(i, ofs, size, w, h);
 			Javelin::RgbaBitmap bm(w, h);
-			copymem(bm.GetData(), &r[ofs], size);
-			{
+			for (unsigned j = 0; j < size / 4; j++) {
 				Javelin::ColorRgba<unsigned char> *dp = bm.GetData();
-				for (int j = 0; j < size / 4; j++) {
-					SWAP(dp[j].r, dp[j].b);
-				}
+				/* red and Green colors are swapped.  */
```
Comment (akien-mga) [inline modules/pvr/texture_loader_pvr.cpp]: Wasn't it red and blue that were swapped?

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**